pandas/core/arrays/datetimes.py

```diff
@@ -239,56 +239,41 @@ def _generate_range(cls, start, end, periods, freq, tz=None,
         start, end, _normalized = _maybe_normalize_endpoints(start, end,
                                                              normalize)
 
-        tz, inferred_tz = _infer_tz_from_endpoints(start, end, tz)
-
-        if hasattr(freq, 'delta') and freq != Day():
-            # sub-Day Tick
-            if inferred_tz is None and tz is not None:
-                # naive dates
-                if start is not None and start.tz is None:
-                    start = start.tz_localize(tz, ambiguous=False)
-
-                if end is not None and end.tz is None:
-                    end = end.tz_localize(tz, ambiguous=False)
-
-            if start and end:
-                if start.tz is None and end.tz is not None:
-                    start = start.tz_localize(end.tz, ambiguous=False)
-
-                if end.tz is None and start.tz is not None:
-                    end = end.tz_localize(start.tz, ambiguous=False)
+        tz, _ = _infer_tz_from_endpoints(start, end, tz)
 
+        # Make sure start and end are timezone localized if:
+        # 1) freq = a Timedelta-like frequency (Tick)
+        # 2) freq = None i.e. generating a linspaced range
+        if isinstance(freq, Tick) or freq is None:
+            localize_args = {'tz': tz, 'ambiguous': False}
```

Comment:
I think its worthile to add freq to ``_maybe_localize_point`` then you don;t need ``localize_args``; it will be inside that function